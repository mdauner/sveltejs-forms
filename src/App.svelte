<script>
  import {
    Form,
    Input,
    Select,
    Choice,
  } from './components/components.module.js';
  import * as yup from 'yup';

  function handleReset() {
    console.log('form has been reset');
  }

  function handleSubmit({ detail: { values, setSubmitting, resetForm } }) {
    setTimeout(() => {
      console.log(values);
      setSubmitting(false);
      resetForm({
        user: { email: 'test@user.com' },
      });
    }, 2000);
  }

  const schema = yup.object().shape({
    user: yup.object().shape({
      email: yup
        .string()
        .required()
        .email(),
      password: yup.string().min(4),
      language: yup.string().required(),
      os: yup
        .array()
        .of(yup.string().required())
        .min(2),
    }),
  });

  const langOptions = [
    { id: 'svelte', title: 'Svelte' },
    { id: 'react', title: 'React' },
    { id: 'angular', title: 'Angular' },
  ];

  const osOptions = [
    { id: 'macos', title: 'macOS' },
    { id: 'linux', title: 'Linux üêß' },
    { id: 'windows', title: 'Windows' },
  ];
</script>

<style lang="scss">
  :global(.sveltejs-forms) {
    background-color: lightgray;
    display: inline-block;
    padding: 1rem;
    .field {
      margin-bottom: 1rem;
      &.error {
        input,
        select {
          border: 1px solid red;
        }
        .message {
          margin-top: 0.2rem;
          color: red;
          font-size: 0.8rem;
        }
      }
    }
  }
</style>

<Form
  {schema}
  validateOnChange={true}
  validateOnBlur={true}
  on:submit={handleSubmit}
  on:reset={handleReset}
  let:isSubmitting
  let:isValid>
  <Input name="user.email" placeholder="Email" />
  <Input
    name="user.password"
    type="password"
    placeholder="Password"
    multiline />
  <Select name="user.language" options={langOptions} />
  <Choice name="user.os" options={osOptions} multiple />

  <button type="reset">Reset</button>
  <button type="submit" disabled={isSubmitting}>Sign in</button>
  The form is valid: {isValid}
</Form>
