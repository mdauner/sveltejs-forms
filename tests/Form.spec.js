import App from './TestApp.svelte';
import { render, fireEvent, wait } from '@testing-library/svelte';
import * as yup from 'yup';

describe('Form', () => {
  it('click on submit button dispatches submit event and sets isSubmitting to true', async () => {
    const { component, getByText } = render(App, {
      props: { onSubmit: jest.fn() },
    });
    const signInButton = getByText('Sign in');

    expect(component.form.$$.ctx.$isSubmitting).toBe(false);
    expect(signInButton).not.toHaveAttribute('disabled');

    await fireEvent.click(signInButton);
    expect(component.onSubmit).toHaveBeenCalledTimes(1);
    expect(signInButton).toHaveAttribute('disabled');
    expect(component.form.$$.ctx.$isSubmitting).toBe(true);
  });

  it('shows error message when schema is defined', async () => {
    const schema = yup.object().shape({
      email: yup
        .string()
        .required()
        .email(),
    });
    const { component, getByPlaceholderText, getByText } = render(App, {
      props: { schema },
    });
    const emailInput = getByPlaceholderText('Email');

    await fireEvent.change(emailInput, {
      target: { value: 'invalid value' },
    });
    await wait(() => {
      expect(getByText('email must be a valid email')).toBeInTheDocument();
    });
    expect(component.form.$$.ctx.$errors).toEqual({
      email: 'email must be a valid email',
    });
  });

  it('registers fields and sets default values', async () => {
    const { component } = render(App);

    expect(component.form.$$.ctx.$values).toEqual({
      email: '',
      language: '',
      os: '',
    });
    expect(component.form.$$.ctx.$touched).toEqual({
      email: false,
      language: false,
      os: false,
    });
    expect(component.form.$$.ctx.$errors).toEqual({
      email: null,
      language: null,
      os: null,
    });
  });

  it('sets initial values', async () => {
    const { component } = render(App, {
      props: { initialValues: { email: 'test@user.com' } },
    });

    expect(component.form.$$.ctx.$values).toEqual({
      email: 'test@user.com',
      language: '',
      os: '',
    });
  });

  it('unregisters when field is removed', async () => {
    const { component, getByText, getByPlaceholderText } = render(App);

    expect(component.form.$$.ctx.$values).not.toHaveProperty('optional');
    let showButton = getByText('Show');
    await showButton.click();
    await wait(() => {
      expect(getByPlaceholderText('Optional')).toBeInTheDocument();
      expect(component.form.$$.ctx.$values).toHaveProperty('optional');
    });

    showButton = getByText('Hide');
    await showButton.click();
    expect(component.form.$$.ctx.$values).not.toHaveProperty('optional');
  });

  it('matches snapshot', async () => {
    const { container } = render(App);
    expect(container.firstChild).toMatchSnapshot();
  });
});
