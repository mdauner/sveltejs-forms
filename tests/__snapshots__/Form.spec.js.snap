// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form matches snapshot 1`] = `
<div>
  <form
    class="sveltejs-forms"
  >
    <div
      class="field"
    >
      <input
        name="email"
        placeholder="Email"
        type="text"
      />
       
    </div>
     
     
    <button>
      Show
    </button>
     
    <div
      class="field"
    >
      <select
        name="language"
      >
        <option
          value=""
        />
        <option
          value="svelte"
        >
          Svelte
        </option>
        <option
          value="react"
        >
          React
        </option>
        <option
          value="angular"
        >
          Angular
        </option>
      </select>
       
    </div>
     
    <div
      class="field"
    >
      <input
        id="macos"
        name="os"
        type="radio"
        value="macos"
      />
       
      <label
        for="macos"
      >
        macOS
      </label>
      
      <input
        id="linux"
        name="os"
        type="radio"
        value="linux"
      />
       
      <label
        for="linux"
      >
        Linux üêß
      </label>
      
      <input
        id="windows"
        name="os"
        type="radio"
        value="windows"
      />
       
      <label
        for="windows"
      >
        Windows
      </label>
      
       
    </div>
     
    <button
      type="submit"
    >
      Sign in
    </button>
  </form>
</div>
`;

exports[`Form onSubmit event returns values, resetForm, setSubmitting 1`] = `
Object {
  "email": "test@user.com",
  "language": "svelte",
  "os": "macos",
}
`;

exports[`Form onSubmit event returns values, resetForm, setSubmitting 2`] = `
Object {
  "email": "test@user.com",
  "language": "svelte",
  "os": "macos",
}
`;

exports[`Form onSubmit event returns values, resetForm, setSubmitting 3`] = `
Object {
  "email": null,
  "language": null,
  "os": null,
}
`;

exports[`Form onSubmit event returns values, resetForm, setSubmitting 4`] = `
Object {
  "email": true,
  "language": true,
  "os": true,
}
`;

exports[`Form onSubmit event returns values, resetForm, setSubmitting 5`] = `
Object {
  "email": "",
  "language": "",
  "os": "",
}
`;

exports[`Form onSubmit event returns values, resetForm, setSubmitting 6`] = `
Object {
  "email": null,
  "language": null,
  "os": null,
}
`;

exports[`Form onSubmit event returns values, resetForm, setSubmitting 7`] = `
Object {
  "email": false,
  "language": false,
  "os": false,
}
`;

exports[`Form shows error message when schema is defined 1`] = `
<div>
  <form
    class="sveltejs-forms"
  >
    <div
      class="field error"
    >
      <input
        name="email"
        placeholder="Email"
        type="text"
      />
       
      <div
        class="message"
      >
        email must be a valid email
      </div>
    </div>
     
     
    <button>
      Show
    </button>
     
    <div
      class="field"
    >
      <select
        name="language"
      >
        <option
          value=""
        />
        <option
          value="svelte"
        >
          Svelte
        </option>
        <option
          value="react"
        >
          React
        </option>
        <option
          value="angular"
        >
          Angular
        </option>
      </select>
       
    </div>
     
    <div
      class="field"
    >
      <input
        id="macos"
        name="os"
        type="radio"
        value="macos"
      />
       
      <label
        for="macos"
      >
        macOS
      </label>
      
      <input
        id="linux"
        name="os"
        type="radio"
        value="linux"
      />
       
      <label
        for="linux"
      >
        Linux üêß
      </label>
      
      <input
        id="windows"
        name="os"
        type="radio"
        value="windows"
      />
       
      <label
        for="windows"
      >
        Windows
      </label>
      
       
    </div>
     
    <button
      type="submit"
    >
      Sign in
    </button>
  </form>
</div>
`;
